src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs

```diff
@@ -148,8 +153,14 @@ public int Count
                 // they are being dequeued. If we ever changed this (e.g. to pool nodes somehow),
                 // we'd need to revisit this implementation.
 
-                for (Node curr = _head; curr != null; curr = curr._next)
+                for (Node curr = _head; ReferenceEquals(curr, Guard) == false;)
                 {
+                    // spin till _next is set
```

Comment:
Frankly, if you're going this route you should probably document more _why_ you're waiting for this to be set, especially since the comment just a few lines previous implies that what you're doing here should be nonsense (and thus, the previous comment should be edited).

One thing I would recommend checking out, if you continue with this, is how [`SpinWait.SpinUntil(...)`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.spinwait.spinuntil?view=netframework-4.7.2#System_Threading_SpinWait_SpinUntil_System_Func_System_Boolean__) affects things.

Reply:
@Clockwork-Muse Let me address the issue you described above in the first commit, and then, I'll focus on documenting it not only in the PR but also in the code.